### PR TITLE
 *: let baseFuncDesc.typeInfer return error instead of panic

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
Before this commit, the following sql will make tidb-server panic.
``` sql
tidb> create table t(a int);
Query OK, 0 rows affected (0.02 sec)

tidb> select std(a) from t;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```

Afer this commit,
``` sql
tidb> select std(a) from t;
ERROR 1105 (HY000): unsupported agg function: std
```

### What is changed and how it works?
Let baseFuncDesc.typeInfer return error instead of panic.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by shenli** [comment]
Good job!

**Comment by XuHuaiyu** [comment]
The explain test changed because of this line: 
https://github.com/pingcap/tidb/pull/10910/files#diff-1a802e13255ca722504fe1b0cc5e34abR321

This line should be `p = a.aggPushDown(p)`.

**Comment by XuHuaiyu** [comment]
Two issues need to be fixed after this commit:
https://github.com/pingcap/tidb/pull/10910/files#diff-ab83e4a8f21f295ae805f30d7e889703R345
and
https://github.com/pingcap/tidb/pull/10910/files#diff-ab83e4a8f21f295ae805f30d7e889703R348

File: cmd/explaintest/r/explain_easy.result [hunk 1]
```diff
@@ -455,10 +455,8 @@ TableReader_7	10000.00	root	data:TableScan_6
 └─TableScan_6	10000.00	cop	table:t1, range:[-inf,+inf], keep order:false, stats:pseudo
 explain select distinct t1.a, t1.b from t1 left outer join t2 on t1.a = t2.a;
 id	count	task	operator info
-StreamAgg_18	8000.00	root	group by:col_2, col_3, funcs:firstrow(col_0), firstrow(col_1)
-└─IndexReader_19	8000.00	root	index:StreamAgg_10
-  └─StreamAgg_10	8000.00	cop	group by:test.t1.a, test.t1.b, funcs:firstrow(test.t1.a), firstrow(test.t1.b)
-    └─IndexScan_17	10000.00	cop	table:t1, index:a, b, range:[NULL,+inf], keep order:true, stats:pseudo
+TableReader_9	10000.00	root	data:TableScan_8
```
Comment (crazycs520) [inline cmd/explaintest/r/explain_easy.result]: Why this plan was changed?

Comment (XuHuaiyu) [reply cmd/explaintest/r/explain_easy.result]: https://github.com/pingcap/tidb/pull/10910#issuecomment-504713988

**Pull request merged**